### PR TITLE
Add ability to use importance matrix for all k-quants

### PR DESCRIPTION
**Pull request by ikawrakow**
TL;DR See title

I see improvements in perplexities for all models that I have tried. The improvement is most significant for low-bit quantization. It decreases with bits-per-weight used, and becomes marginal for `Q6_K`.

As an example, the table below compares perplexities for Mistral-7B between master and this PR. The `QError` column is defined as `PPL(Q)/PPL(fp16)-1`, where `PPL(fp16) = 5.6924` for Mistral-7B.

| Qunatization | PPL (Master) | PPL (PR) | QError (Master) | QError (PR) | QError(PR)/Qerror(Master) |
|--:|--:|--:|--:|--:|--:|
|Q2_K     |  6.7580 | 6.4268 | 18.72% | 12.90% | 0.689 |
|Q3_K_S |  6.0692 | 6.0022 | 6.62% | 5.44% | 0.822 |
|Q3_K_M | 5.8894 | 5.8490 | 3.46%  | 2.75% | 0.795 |
|Q4_K_S | 5.7764 | 5.7428 | 1.48% | 0.89% | 0.601 |
|Q4_K_M | 5.7539 | 5.7353 | 1.08% | 0.75% | 0.694 |
|Q5_K_S | 5.7258 | 5.7107 | 0.59% | 0.32% | 0.542 |
|Q5_K_M | 5.7168 | 5.7097 | 0.53% | 0.30% | 0.566 |
|Q6_K      | 5.7028 | 5.7011 | 0.18% | 0.15% | 0.833 |

Note that if no importance matrix is supplied to the `quantize` tool, quantization will be identical to what we had before this PR.

### CODE REVIEW

**Comment by JohannesGaessler** [comment]
Sorry, I'm not up-to-date with all of your recent developments. The importance matrix is determined from some corpus of text, correct? If so, which text was used for the importance matrix?

**Comment by ikawrakow** [comment]
> Sorry, I'm not up-to-date with all of your recent developments. The importance matrix is determined from some corpus of text, correct? If so, which text was used for the importance matrix?

Yes, one needs an inference run with the `imatrix` tool that was added in PR #4861 using some corpus of training text. This creates an "importance matrix" that is then used in the `quantize` too via the new `--imatrix` command line argument. For the above results I have used 50,000 tokens from `wiki.train.raw`, but everyone is free to use whatever training data they believe is best.

**Review by ggerganov — APPROVED** [review]
@ikawrakow Thank you for these great contributions and analysis!

**Comment by ikawrakow** [comment]
@Ttl 

Now that all bits and pieces needed for the improved quantization have been added to `llama.cpp`, I'm curious to see a comparison between the KL-divergence of `k-quants` that utilize an importance matrix and what you had done in PR #4739

**Comment by Ttl** [comment]
Some tests with phi-2 Q4_K_S:

<details><summary>Results</summary>

```
Model: phi-2-q4_k_s_imatrix_master.gguf
Size: 1.52 GiB, (BPW 4.68)
Tokens: 286799
KL-divergence:
mean: 0.0358178, [0.0355851 - 0.0360505]
q90: 0.07286, [0.07242 - 0.0733]
q95: 0.1044, [0.1035 - 0.1052]
q99: 0.2238, [0.2201 - 0.2278]
max: 2.887
Reference top token in eval top-n probability:
ref_top1: 0.899 ± 0.00145
ref_top5: 0.9977 ± 0.0002319
ref_top10: 0.99954 ± 0.0001032
Eval top token in reference top-n probability:
eval_top5: 0.9976 ± 0.0002348
eval_top10: 0.999578 ± 9.877e-05

Model: phi-2-q4_k_s.gguf (Previous master)
Size: 1.5 GiB, (BPW 4.64)
Tokens: 286799
KL-divergence:
mean: 0.0419483, [0.0416933 - 0.0422033]
q90: 0.08505, [0.08456 - 0.08553]
q95: 0.1193, [0.1184 - 0.1203]
q99: 0.2509, [0.2472 - 0.2547]
max: 1.998
Reference top token in eval top-n probability:
ref_top1: 0.8894 ± 0.001508
ref_top5: 0.997 ± 0.0002625
ref_top10: 0.999425 ± 0.0001153
Eval top token in reference top-n probability:
eval_top5: 0.9969 ± 0.0002653
eval_top10: 0.999351 ± 0.0001224

Model: phi-2-q4_k_s_pr.gguf (PR4739)
Size: 1.5 GiB, (BPW 4.64)
Tokens: 286799
KL-divergence:
mean: 0.0377337, [0.0375068 - 0.0379605]
q90: 0.0766, [0.07619 - 0.07701]
q95: 0.1076, [0.1068 - 0.1085]
q99: 0.2237, [0.2199 - 0.2274]
max: 2.39
Reference top token in eval top-n probability:
ref_top1: 0.8947 ± 0.001476
ref_top5: 0.9976 ± 0.0002338
ref_top10: 0.999582 ± 9.837e-05
Eval top token in reference top-n probability:
eval_top5: 0.9973 ± 0.0002503
eval_top10: 0.999491 ± 0.0001085

q90, q95, q99 are KL divergence 90%, 95% and 99% quantiles. Reference is fp16 model.
```
</details>

Imatrix was calculated from `wiki.train` for 250 chunks. KL divergence and top token agreement is now better than previously and it's also better than in #4739. File size is slightly larger 4.68 vs 4.64 BPW. I didn't requantize older models, they are the same files as before. KL divergence results are slightly different than in the #4739, I swapped the argument order so that it now calculates Dkl(fp16 | q) instead of the opposite previously the old models are re-evaluated with this change.

Perplexity is slightly better 10.9840 vs 11.0052 before.

**Comment by kalomaze** [comment]
[mistral_importance_nonsense.zip](https://github.com/ggerganov/llama.cpp/files/13933621/mistral_importance_nonsense.zip)
I've calculated a Mistral 7b importance matrix using a bunch of high entropy, "nonsense" generations from the model itself (in the hopes to maximize any "outlier" activations and avoid overfitting the importance matrix.)

image

I will try to test the quant PPL with this

**Comment by kalomaze** [comment]
I'm having trouble telling whether or not it's actually having an impact, there doesn't appear to be any actual debugging logs or print statements during quantization that would point to the importance matrix being used, and the outputs appear the same for 4_K_S.

**Comment by kalomaze** [comment]
image
Mistral running in less than 1GB of RAM, and _still_ mustering up [somewhat] coherent text despite the overwhelming brain damage it faces. Impressive.

**Comment by sorasoras** [comment]
@ikawrakow I find that's strange I got better ppl with Q3MK  vs  Q4MK.

**Comment by askmyteapot** [comment]
So i did a mixtral Q3KS (modified slightly to have Q6K Output and K weights), calculated the imp.matrix on Q8_0 (as full 16bit was going to take 100hrs on my 64GB DDR4 system). Took about 10 hours to complete. 

I used the wiki.test.raw file instead of the training one, but for the full run. 

Results are:

Final estimate: PPL = 4.6058 +/- 0.02510 - mod3ks with imatrix
Final estimate: PPL = 4.6288 +/- 0.02509 - mod3ks

is it worth 10hrs of CPU processing.... hmm.. not sure.

**Comment by ikawrakow** [comment]
> @ikawrakow I find that's strange I got better ppl with Q3MK vs Q4MK.

@sorasoras Can you share more details?

**Comment by ikawrakow** [comment]
@askmyteapot Thanks for the observations. Three comments:

1. Never ever use data utilized for testing to train any part of your model. Not even the importance matrix
2. The PPL improvement due to the importance matrix for Mistral-Instruct-8x7B and `Q3_K_S` is modest. I have a 16 GB GPU, so `Q3_K_S` of Mixtral-8x7B cannot be fully loaded. For 24 layers offloaded to the GPU, I get `PPL(no imatrix) = 4.6677` and `PPL(imatrix) = 4.6283`. We have `PPL(fp16) =  4.4091`, so quantization error is 5.86% without, and 4.97% with imatrix. If you think that 18% reduction in quantization error is not worth the effort, you don't need to use it
3. At https://huggingface.co/datasets/ikawrakow/imatrix-from-wiki-train/tree/main you can find ready imatrix data for a bunch of models. The one for Mixtral-Instruct-8x7B is 25.7 MB, which would have taken much less than a minute to download, and would have saved you the 10 hours compute.

**Comment by sorasoras** [comment]
> > @ikawrakow I find that's strange I got better ppl with Q3MK vs Q4MK.
> 
> @sorasoras Can you share more details?

.\sakura0.9_Q3KM_ix_wikitest.gguf -f .\wiki.test.raw   PPL = 10.4118 +/- 0.07169
.\sakura0.9_Q4KM_ix_wikitest.gguf -f .\wiki.test.raw    PPL = 10.4238 +/- 0.07183
both of them get inimax via wiki test, but also other set of Chinese/Japanese data

by the way, When I use q2xs, the output just broken down. Does it has to do with compile with blas, openblas specific

**Comment by ikawrakow** [comment]
One more thing related to @askmyteapot's comment: I cannot confirm with my setup a perplexity of `PPL(no imatrix) = 4.6288` with @askmyteapot's `Q3_K_S` modifications (`attn_output` tensor quantized with `Q6_K` instead of `Q5_K`, `attn_k` tensors with `Q6_K` instead of `Q8_0`). With these changes I get `PPL(no imatrix) = 4.6666`, so basically the same as without the changes. Which means that either
* @askmyteapot has additional changes that lower the perplexity
* Partial offload is having an unexpectedly large impact on perplexity. This could be because of PR #4766 (see [my comment there](https://github.com/ggerganov/llama.cpp/pull/4766#issuecomment-1890363264)), or because of a significant loss of accuracy when running on the CPU (see issue #4755).
* Perhaps these are `Q3_K_M` models rather than `Q3_K_S`? But for `Q3_K_M` and no imatrix I get `PPL = 4.6141`, so lower than the value @askmyteapot reports.

**Comment by ikawrakow** [comment]
@sorasoras I still don't know what model you are using. https://huggingface.co/kyujinpy/Sakura-SOLAR-Instruct ?

**Comment by sorasoras** [comment]
> @sorasoras I still don't know what model you are using. https://huggingface.co/kyujinpy/Sakura-SOLAR-Instruct ?

https://huggingface.co/SakuraLLM/Sakura-13B-LNovel-v0.9-GGUF
It's a qwen finetune specific for lightnovel translation：）

**Comment by ikawrakow** [comment]
@sorasoras The model I downloaded from your link has a `wiki.test.raw` `fp16` perplexity of 10.6003. This is pretty high for a 14B model. You are getting `PPL = 10.4118` with `Q3_K_M`. Are we talking about the same model? But it gets worse: the `fp16` perplexity on your calibration data is 482.9229 (!!!), so the model basically does not know what to do with this text. Not sure it is helping to guide the quantization. With `Q6_K` quantization, which is usually within 0.15% of `fp16`, I get `PPL = 496.7347` for `calib_jp_to_zh.raw`, so almost 3% higher. This tells me that k-quants are somehow not working very well for this model. All this is before even thinking about the importance matrix.

**Comment by sorasoras** [comment]
> @sorasoras The model I downloaded from your link has a `wiki.test.raw` `fp16` perplexity of 10.6003. This is pretty high for a 14B model. You are getting `PPL = 10.4118` with `Q3_K_M`. Are we talking about the same model? But it gets worse: the `fp16` perplexity on your calibration data is 482.9229 (!!!), so the model basically does not know what to do with this text. Not sure it is helping to guide the quantization. With `Q6_K` quantization, which is usually within 0.15% of `fp16`, I get `PPL = 496.7347` for `calib_jp_to_zh.raw`, so almost 3% higher. This tells me that k-quants are somehow not working very well for this model. All this is before even thinking about the importance matrix.

I am getting 10.1793 with FP16
with FP16 , I need to partially offload with rocm. I don't know if that would affect .
wiki.test perplexity is high and that's expected as this model is used to do translation exclusively from Japanese to Chinese, so other ability like english is sacrifice during finetune.

as for my dataset, when your dataset is  multilingual, it would get ridiculous high. It contains many pair of two languages, Chinese and Japanese one to one pair, one book in japanese to official chinese version.

I would say imatrix does help Q2KM and Q3KL to get a more consistent and higher quality output compare to the non-imatrix one, but i dont think it's a good idea to compute perplexity with that data as the perplexity output is just not very stable.

Q2XS just output garbage but Q2KM does work.

that been said, I still want to know if there is other reason Q4KM is the same as Q3KM

by using  wiki.test.raw to calculate imatrix, and use that to generate quantization and use that to calculate perplexity.
maybe I need a more suitable dataset to get better results? I am not so sure.

As for K-quant,maybe I should give Q5_1 Q4_1 a try xd and wait for that imatrix for "legacy quantization"

**Comment by askmyteapot** [comment]
> One more thing related to @askmyteapot's comment: I cannot confirm with my setup a perplexity of `PPL(no imatrix) = 4.6288`

I was a bit tired when I typed that. The q8 k is the same, the v and output are q6k. I had to add in an extra if statement to find the attn_v. All weights are q3. I was trying to find the biggest I could go with 256batch and 12288ctx in 24gb (P40). 

Also, if we are just getting activations for the matrix, wouldn't testing data be better than training data?

**Comment by ikawrakow** [comment]
> Also, if we are just getting activations for the matrix, wouldn't testing data be better than training data?

No, because your goal is not to minimize the score of the test set, but to get a good quantization for the model. Hence, you train on a training dataset and test on a test dataset that was never involved in the training in any way.

**Comment by ikawrakow** [comment]
> @ikawrakow is there a way to benchmark the quantized gguf against mmlu, gsm8k, and etc?

No, the only metric that `llama.cpp` can currently compute that is commonly used for LLM evaluation is HellaSwag.

I do agree that it would be great if one could run these tests directly with `llama.cpp`.

**Comment by sorasoras** [comment]
> > @ikawrakow is there a way to benchmark the quantized gguf against mmlu, gsm8k, and etc?
> 
> No, the only metric that `llama.cpp` can currently compute that is commonly used for LLM evaluation is HellaSwag.
> 
> I do agree that it would be great if one could run these tests directly with `llama.cpp`.

```
.\sakura0.9_Q2_K_imx.gguf   -f .\calib_jp_to_zh.raw PPL = 613.2313 +/- 13.09685
.\sakura0.9_Q3_K_M_imx.gguf -f .\calib_jp_to_zh.raw PPL = 607.2495 +/- 13.37079
.\sakura0.9_Q4_K_M_imx.gguf -f .\calib_jp_to_zh.raw PPL = 608.0662 +/- 13.42431
.\sakura0.9_Q5_K_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 569.2346 +/- 12.27382
.\sakura0.9_Q6_K_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 542.1290 +/- 11.65621

.\sakura0.9_Q4_0_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 573.9627 +/- 12.46049  PPL = 573.9627 +/- 12.46049
.\sakura0.9_Q4_1_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 611.1743 +/- 13.16429
.\sakura0.9_Q5_0_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 601.3890 +/- 13.24175
.\sakura0.9_Q5_1_imx.gguf -f .\calib_jp_to_zh.raw   PPL = 618.0934 +/- 13.50312
```

I think you‘re right k-quant. it does seems have problem with Kquat.

The best option for me could be AWQ+Q4_0 lol

**Comment by JianbangZ** [comment]
> > @ikawrakow is there a way to benchmark the quantized gguf against mmlu, gsm8k, and etc?
> 
> No, the only metric that `llama.cpp` can currently compute that is commonly used for LLM evaluation is HellaSwag.
> 
> I do agree that it would be great if one could run these tests directly with `llama.cpp`.

I figured out a way to test gguf accuracy with https://github.com/EleutherAI/lm-evaluation-harness, by leveraging https://github.com/abetlen/llama-cpp-python?tab=readme-ov-file#openai-compatible-web-server to start a local host server. 
I am able to run mmlu, c-eval, gsm8k and etc. However, it is extremely slow.
